src/NHibernate.Test/CacheTest/FilterKeyFixture.cs

```diff
@@ -8,105 +9,129 @@ namespace NHibernate.Test.CacheTest
+		private void FilterDescLikeToCompare(out FilterKey fk, out FilterKey fk1, bool sameValue)
 		{
 			const string filterName = "DescriptionLike";
 			var f = new FilterImpl(Sfi.GetFilterDefinition(filterName));
 			f.SetParameter("pLike", "so%");
-			fk = new FilterKey(filterName, f.Parameters, f.FilterDefinition.ParameterTypes);
+			fk = new FilterKey(f);
 
 			var f1 = new FilterImpl(Sfi.GetFilterDefinition(filterName));
-			f1.SetParameter("pLike", "%ing");
-			fk1 = new FilterKey(filterName, f.Parameters, f.FilterDefinition.ParameterTypes);
+			f1.SetParameter("pLike", sameValue ? "so%" : "%ing");
+			fk1 = new FilterKey(f1);
 		}
 
-		private void FilterDescValueToCompare(out FilterKey fk, out FilterKey fk1)
+		private void FilterDescValueToCompare(out FilterKey fk, out FilterKey fk1, bool sameValue)
 		{
 			const string filterName = "DescriptionEqualAndValueGT";
 			var f = new FilterImpl(Sfi.GetFilterDefinition(filterName));
 			f.SetParameter("pDesc", "something").SetParameter("pValue", 10);
-			fk = new FilterKey(filterName, f.Parameters, f.FilterDefinition.ParameterTypes);
+			fk = new FilterKey(f);
+
+			var f1 = new FilterImpl(Sfi.GetFilterDefinition(filterName));
+			f1.SetParameter("pDesc", "something").SetParameter("pValue", sameValue ? 10 : 11);
+			fk1 = new FilterKey(f1);
+		}
+
+		private void FilterValueInToCompare(out FilterKey fk, out FilterKey fk1, bool sameValue)
+		{
+			const string filterName = "ValueIn";
+			var f = new FilterImpl(Sfi.GetFilterDefinition(filterName));
+			f.SetParameterList("pIn", new HashSet<int> { 10, 11 });
+			fk = new FilterKey(f);
 
 			var f1 = new FilterImpl(Sfi.GetFilterDefinition(filterName));
-			f1.SetParameter("pDesc", "something").SetParameter("pValue", 11);
-			fk1 = new FilterKey(filterName, f.Parameters, f.FilterDefinition.ParameterTypes);
+			f1.SetParameterList("pIn", sameValue ? (ICollection<int>)new [] { 10, 11 } : new HashSet<int> { 10, 12 });
+			fk1 = new FilterKey(f1);
 		}
 
 		[Test]
 		public void NotEquality()
 		{
-			FilterKey fk, fk1;
-			FilterDescLikeToCompare(out fk, out fk1);
+			FilterDescLikeToCompare(out var fk, out var fk1, false);
+			Assert.That(fk, Is.Not.EqualTo(fk1), "fk & fk1");
 
-			FilterKey fvk, fvk1;
-			FilterDescValueToCompare(out fvk, out fvk1);
+			FilterDescValueToCompare(out var fvk, out var fvk1, false);
+			Assert.That(fvk, Is.Not.EqualTo(fvk1), "fvk & fvk1");
 
-			Assert.That(fk, Is.Not.EqualTo(fvk));
-			Assert.That(fk1, Is.Not.EqualTo(fvk1));
+			FilterValueInToCompare(out var fik, out var fik1, false);
+			Assert.That(fik, Is.Not.EqualTo(fik1), "fik & fik1");
+
+			Assert.That(fk, Is.Not.EqualTo(fvk), "fk & fvk");
+			Assert.That(fk1, Is.Not.EqualTo(fvk1), "fk1 & fvk1");
+			Assert.That(fvk, Is.Not.EqualTo(fik), "fvk & fik");
+			Assert.That(fvk1, Is.Not.EqualTo(fik1), "fvk1 & fik1");
 		}
 
 		[Test]
 		public void HashCode()
 		{
-			// HashCode is aware only by parameters names not values (should work as Equal)
```

Comment:
Unfortunately true for the current hashcode implementation, which is bad for performances. But in fact it should ideally not be equal, since actually filter with different parameters values are not equals.